### PR TITLE
Made the tls configuration injectable

### PR DESCRIPTION
**Pull request by Loupax**
Had to rewrite the `Send` function so that we can allow the user to inject their own `tls.Config`

Feel free to provide feedback if you want :)

### CODE REVIEW

**Comment by jazzboME** [comment]
I don't think it's a great idea to change the .New() constructor?   This will potentially break a lot of code when people update.

Why not add something like .NewWithTLS()?  See comments in https://github.com/domodwyer/mailyak/issues/36

**Comment by Loupax** [comment]
Sorry for the unsolicited PR, I kinda opened it by accident :P

I like the idea of the `.NewWithTLS()` factory, should give it a try when I
get back to a PC.

Στις Παρ, 20 Σεπ 2019 στις 7:17 μ.μ., ο/η Kevin Foss <
notifications@github.com> έγραψε:

**Comment by domodwyer** [comment]
Hi @Loupax, @jazzboME 

Always happy to receive a PR, including the unsolicited ones!

Thanks for taking the time to open this, though @jazzboME makes a good point about the constructor. I would love to merge this with the proposed `NewWithTLS()` constructor :)

Thank you both,
Dom

**Comment by Loupax** [comment]
Looks like it's going to need some more work to ensure exact behaviour.

The way I see it, we need to copy the SDL smtp functionality into the package. It marked as frozen by the Go team so we should not worry about future changes but you can never be 100% sure.

Are you people ok with with the entire `smtp` package becoming part of mailyak?

**Comment by Loupax** [comment]
Here is a functional PR, if it needs more polishing or if you find any issues please let me know.

What happened here is the following:

1. Copied the `smtp` package from the standard library. This had to be done so we can add the tls configuration with no unintentional side effects
2. Made any functions inside `smtp.go` file private so we don't expose irrelevant contracts that we have to maintain
3. Created new factory function `NewWithTLS()` to `mailyak` so that the user can create a `Client` with custom TLS

**Comment by domodwyer** [comment]
Hi @Loupax 

Thanks for the PR :)

I was just wondering, why did we have to copy the smtp code? I see you say it was to avoid "side affects" but could you elaborate?

Thanks again!
Dom

**Comment by Loupax** [comment]
Hello @domodwyer 

Yes, I guessed the MR would look scary 😄 

TL;DR: When I said side effects, I basically meant "unintentional breaking changes"

Everything started when I realized that the original MR did much fewer things than the original functionality that used the native ` smtp.SendMail` function. 
For example the original `smtp.SendMail` would return an error for invalid email addresses, deal with handshaking with the SMTP server and made checks about whether the server supports AUTH or not. 

_This is when I realized that the original MR wouldn't guarantee 100% same behavior with the existing functionality_

Initially I started moving functionality piece by piece to ensure we maintain backwards compatibility for existing users with as few changes as possible, and ended up copying **almost** the entire `smtp` package. You can see that all structs that are used by the `smtp` package are still the SDL ones. 

I really don't like adding a huge file in the project but this is the best I could come up with, maybe we can figure out something better together 😄 Maybe put the extra file in the `internal` package so we don't have to provide support to consumers?

**Comment by domodwyer** [comment]
Hi @Loupax 

Sorry for the delay - I wanted to take a proper look at this and I've been away for a bit.

I diff-ed the [smtp.go file in this PR] and the [stdlib smtp.go] and aside from un-exporting the methods (I assume so we don't pollute the mailyak API - nice!) the only change I could see was adding `config *tls.Config` to `sendMail()`, however it appears unused?

If the intention was to use it in the [`STARTTLS` conditional branch] it is shadowed immediately by the `config := ...` declaration:

```go
	if ok, _ := c.Extension("STARTTLS"); ok {
		config := &tls.Config{ServerName: c.serverName}
		if testHookStartTLS != nil {
			testHookStartTLS(config)
		}
		if err = c.StartTLS(config); err != nil {
			return err
		}
	}
```

Unless I'm missing something, I'm not sure if this PR functions as intended - I think the best way to add this would be to not use the `SendMail()` function, but instead dial and manage the connection ourselves - this would also avoid having to pull in the stdlib SMTP implementation.

Let me know if I've overlooked something, happy to help :)

Dom

[smtp.go file in this PR]: https://github.com/domodwyer/mailyak/blob/b88691bb0b9c4158e6b3ecf0f9c488079b777f9e/smtp.go
[stdlib smtp.go]: https://golang.org/src/net/smtp/smtp.go
[`STARTTLS` conditional branch]: https://golang.org/src/net/smtp/smtp.go?s=9230:9308#L336

**Comment by Loupax** [comment]
About the config, yes you are right, the existing code needs to be corrected, thanks for the feedback.

Now about whether or not we should handle the `sendMail` ourselves, I agree it will end up being better than using copy-pasted code, but unfortunately I can't guarantee I'll have the time to do this properly any time soon...

I can close this MR for now, so that everyone knows this is still pending and I can reopen it if I get the chance to work on it again :)
